syntax = "proto3";

message GPSPacket {
    double lat = 1;
    double lng = 2;
    double alt = 3;
    double speed = 4;
    double course = 5;
    uint32 satellites = 6;
}

message TemperaturePacket {
    float coolant = 1;
    float air = 2;
    float fuel = 3;
    float oil = 4;
}

message SwitchesPacket {
    bool neutral = 1;
    bool oil_pressure_light = 2;
    bool launch_control_active = 3;
    bool launch_control_switch = 4;
    bool anti_lag_switch = 5;
    bool thermo_fan = 6;
    float rotary_trim_pot_1 = 7;
    float rotary_trim_pot_2 = 8;
    float rotary_trim_pot_3 = 9;
    bool check_engine_light = 10;
    bool pit_lane_speed_limiter_active = 11;
    bool pit_lane_speed_limiter_switch_state = 12;
}

message BrakesPacket {
    float pressure_front = 1;
    float pressure_rear = 2;
    float pressure_front_ratio = 3;
    float pressure_rear_ratio = 4;
    float pressure_difference = 5;
}

message GForcePacket {
    float lateral_g = 1;
    float longitudinal_g = 2;
    float vertical_g = 3;
}

message DamperPacket {
    float travel_front_left = 1;
    float travel_rear_left = 2;
    float travel_front_right = 3;
    float travel_rear_right = 4;
}

message RatePacket {
    float pitch_rate = 1;
    float roll_rate = 2;
    float yaw_rate = 3;
}

message SuspensionPacket {
    BrakesPacket brakes = 1;
    GForcePacket gforce = 2;
    DamperPacket damper = 3;
    RatePacket rate = 4;
}

message EnginePacket {
    uint32 rpm = 1;
    float manifold_pressure = 2;
    float throttle_position = 3;
    float fuel_pressure = 5;
    float oil_pressure = 6;
    float engine_demand = 7;
    float sensor_1 = 8;
    float sensor_2 = 9;
    float overall = 10;
    float vehicle_speed = 11;
    float battery_voltage = 12;
    float barometric_pressure = 13;
    TemperaturePacket temperature = 14;
    SwitchesPacket switches = 15;
    float steering_angle = 16;
    uint32 gear = 17;
    float accelerator_pedal_position = 18;
}

message TelemetryPacket {
    uint32 timestamp = 1;
    float voltage = 2;
    bool can_connected = 3;
    uint32 can_messages = 4;
    bool gps_valid = 5;
    GPSPacket gps = 6;
    EnginePacket engine = 7;
    SuspensionPacket suspension = 8;
}